### PR TITLE
Feat: Adiciona compatibilidade com plugins

### PR DESCRIPTION
**Pull request by JoaoPedro0000**
Com esse PR o plugin passa a ter compatibilidade com kits e grupos de produtos (WPC Composite e WPC Product Bundles)

### CODE REVIEW

File: Services/Products/ProductsService.php [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+	{
+		return $product->get_type() === self::COMPOSITE_TYPE ||
+			$product->get_type() === self::PRODUCT_COMPOSITE_TYPE;
+	}
+
+	public static function isBundleProduct($product): bool
+	{
+		return $product->get_type() === self::BUNDLE_TYPE ||
+			$product->get_type() === self::PRODUCT_BUNDLE_TYPE;
+	}
```
Comment (tsantos8080) [inline Services/Products/ProductsService.php:29]: Acredito que isso estar nessa classe não faz muito sentido, pois já tem uma classe pro Composite e Bundle, daí os conceitos desses dois também estão aqui.

Não daria pra mover esses métodos pra dentro das suas classes? Tipo, lá dentro do `CompositeService` ter um:
```php
public function isProductAggregate($product): bool
{
    return $product->get_type() === self::COMPOSITE_TYPE ||
			$product->get_type() === self::PRODUCT_COMPOSITE_TYPE;
}
```

Dentro do `BundleService` também e dentro do `ProductService` também, mas este último retornando sempre `false`, só pra manter a mesma interface.
<br/>

Daí dessa forma, pra utilizar esse cara no resto do código, você poderia usar assim:
```php
    $productService = ProductServiceFactory::create($product);

    if ($productService->isProductAggregate($product)) {
      // ...
    }
```

Acredito que dessa forma ficaria mais fácil pra lidar com as regras, já que o chamador do `ProductService` só precisaria verificar se o produto é agregado ou não, sem saber das regras do plugin.

O que acha?

File: Services/Products/ProductsService.php [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+			if (! empty($product->pricing) &&
+				! empty($product->shipping_fee) &&
+				( $product->pricing == 'include' || $product->pricing == 'only')
+				&& $product->shipping_fee == 'each'
+			) {
+				$valueBase += wc_get_product($product->id)->get_price();
+			}
+		}
+		return $valueBase;
+	}
```
Comment (tsantos8080) [inline Services/Products/ProductsService.php:44]: Parece que essa função é código morto. O que acha de remover?

File: Services/Products/ProductsService.php [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+	/**
+	 * function to remove the price field from
+	 * the product to perform the quote without insurance value
+	 *
+	 * @param array $products
+	 * @return array
+	 */
+	public function removePrice( $products ) {
```
Comment (tsantos8080) [inline Services/Products/ProductsService.php:90]: Essa função tá sendo usada no `PayloadService` mas eu tenho 99% de certeza que remover o `products.*.insurance_value` não tem utilidade nenhuma, já que o carrinho só considera o valor de `options.insurance_value`. Logo, acredito que isso é código quase morto, que só serve pra adicionar complexidade.

Talvez seria útil criar uma tarefa pra ver esse cara posteriormente, ou pelo menos refatorar isso aqui porquê o nome da função tá dizendo que remove o `price`, mas não tem `price` no objeto. Já o comentário diz que remove o "insurance value"... tá muito sem sentido isso :confused:

File: Services/Products/ProductsService.php [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+	/**
+	 * @param object $product
+	 * @return bool
+	 */
+	private function isObjectProduct( $item ) {
+		return (
+			! is_array( $item ) &&
+			(
+				get_class( $item ) == WooCommerceBundleProductsService::OBJECT_PRODUCT_SIMPLE ||
+				get_class( $item ) == WooCommerceBundleProductsService::OBJECT_WOOCOMMERCE_BUNDLE
```
Comment (tsantos8080) [inline Services/Products/ProductsService.php:169]: Isso aqui não deveria ter também a verificação do composite? Não entendi muito bem se está verificando o plugin que gerou o produto :thinking:

File: Services/Products/ProductsService.php [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+		return $data;
+	}
+	/**
+	 * @param $product
+	 * @param $price
+	 * @param int $quantity
+	 * @return Product
+	 */
+	public function normalize($product, $price, $quantity = 1): Product
```
Comment (tsantos8080) [inline Services/Products/ProductsService.php:217]: Fiquei um pouco confuso com esse nome `$product`. No caso, isso seria um daqueles itens dentro do produto, o agregado de produtos, certo?

Um nome diferente de `product` não ficaria mais fácil de entender? Minha sugestão é `productAggregate` pra poder ficar diferente de `compositeProduct` ou `bundleProduct`, o que causaria confusão com os plugins.

```suggestion
	public function normalize($productAggregate, $price, $quantity = 1): Product
```

O que acha?

File: Services/Products/ProductsService.php [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+				);
+			}
+		}
+
+		if ( ! empty( $title ) ) {
+			$title = substr( $title, 0, -2 );
+		}
+
+		return 'Produto(s): ' . $title;
+	}
```
Comment (tsantos8080) [inline Services/Products/ProductsService.php:286]: Isso aqui também parece ser código morto :confused:

File: Services/QuotationService.php [hunk 7]
```diff
@@ -102,6 +99,21 @@ public function calculateQuotationByPostId( $postId ) {
 		return ( new OrderQuotationService() )->saveQuotation( $postId, $quotations );
 	}
 
+	public function getPayload($postId, $products = array())
+	{
+		if (empty($products)) {
+			$productService = new OrdersProductsService();
+			$products = $productService->getProductsOrder($postId);
```
Comment (tsantos8080) [inline Services/QuotationService.php:106]: Só pra não confundir na leitura, o que acha de:
```suggestion
			$orderProductService = new OrdersProductsService();
			$products = $orderProductService->getProductsOrder($postId);
```

**Review by tsantos8080 — APPROVED** [review]

File: Controllers/ShowCalculatorProductPage.php [hunk 8]
```diff
@@ -100,6 +101,14 @@ public function addCalculateShipping() {
...
...return "<div id='woocommerce-correios-calculo-de-frete-na-pagina-do-produto' class='containerCalculator'> <?php wp_nonce_field('solicita_calculo_frete', 'solicita_calculo_frete'); ?> <input type='hidden' id='calculo_frete_endpoint_url' value='".admin_url( 'admin-ajax.php' )."'> <input type='hidden' id='calculo_frete_produto_altura' value='".$this->height."'> <input type='hidden' id='calculo_frete_produto_largura' value='".$this->width."'> <input type='hidden' id='calculo_frete_produto_comprimento' value='".$this->length."'> <input type='hidden' id='calculo_frete_produto_peso' value='".$this->weight."'> <input type='hidden' id='calculo_frete_produto_preco' value='".$this->price."'> <input type='hidden' id='id_produto' value='".$this->id."'> <div class='calculatorRow'> <div class='row'> <div class='col-75'> <p>Simulação de frete</p>". $this->verifyTypeProduct() . "</div></div><div id='calcular-frete-loader'> <img src='".MELHORENVIO_ASSETS."/images/loader.gif' /> </div> <div class=resultado-frete tableResult> <table> <thead> </thead> <tbody> </tbody> </table> <small id='destiny-shipping-mehor-envio'></small></br> <small class='observation-shipping-free'></small> </div> </div> </div>";
```
[the remark this answers is not in the document]
Comment (JoaoPedro0000) [reply Controllers/ShowCalculatorProductPage.php:104]: Eu acho que faz sentido sim, porém não acha que isso faz mais sentido sendo um componente do front realmente? ao invés de uma função que retorna um html? vou deixar assim por enquanto e deixar como um débito técnico a criação de um componente, o que acha?

File: Controllers/ShowCalculatorProductPage.php [hunk 9]
```diff
@@ -100,6 +101,14 @@ public function addCalculateShipping() {
...
+	}
+
+	private function verifyTypeProduct(){
```
[the remark this answers is not in the document]
Comment (tsantos8080) [reply Controllers/ShowCalculatorProductPage.php:107]: Eu não gosto muito porquê não deixa claro que o método retorna um html, mas já é melhor

File: Services/CalculateShippingMethodService.php [hunk 10]
```diff
@@ -38,19 +38,18 @@ class CalculateShippingMethodService {
 	 * @param int    $timeExtra
 	 * @param int    $percent
 	 * @return bool
+	 * @throws Exception
 	 */
-	public function calculateShipping( $package = array(), $code, $id, $company, $title, $taxExtra, $timeExtra, $percent ) {
+	public function calculateShipping( $package, $code, $id, $company, $title, $taxExtra, $timeExtra, $percent ) {
```
[the remark this answers is not in the document]
Comment (JoaoPedro0000) [reply Services/CalculateShippingMethodService.php:43]: Não tem problema, o package mesmo se não tiver nada ele vem como array então não tem problema

File: Services/CalculateShippingMethodService.php [hunk 11]
```diff
@@ -38,19 +38,18 @@ class CalculateShippingMethodService {
...
 		$to = PostalCodeHelper::postalcode( $package['destination']['postcode'] );
 		if ( strlen( $to ) != PostalCodeHelper::SIZE_POSTAL_CODE ) {
 			return false;
 		}
 
-		$products = ( isset( $package['contents'] ) )
-			? $package['contents']
-			: ( new CartWooCommerceService() )->getProducts();
+		$products = ( new CartWooCommerceService() )->getProducts();
```
[the remark this answers is not in the document]
Comment (JoaoPedro0000) [reply Services/CalculateShippingMethodService.php:49]: Isso exatamente, ele vai funcionar como um fallback mesmo, na maior parte dos casos não será usado.

File: Services/CartWooCommerceService.php [hunk 12]
```diff
@@ -23,28 +27,17 @@ public function getProducts() {
...
+			$productService = ProductFactory::createProductServiceById($productId);
 
-			if ( empty( $productInfo ) ) {
-				continue;
+			$products[] = $productService->getDataByProductCart($itemProduct, $items);
+		}
+
+		foreach ($products as $key => $product) {
+			if (isset($product->parentId)){
+				unset($products[$key]);
```
[the remark this answers is not in the document]
Comment (JoaoPedro0000) [reply Services/CartWooCommerceService.php:37]: Na verdade eu queria transformar isso em uma função separada algo como `removeComponentProducts`, o que acha?

File: Services/Products/CompositeService.php [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+						$items[$key]['line_total'] / $items[$key]['quantity'],
+						$items[$key]['quantity']);
+				}
+
+				if (($data->pricing == 'include' || $data->pricing == 'only') && $data->shipping_fee == 'each') {
+					array_filter($data->components, function($component) {
+						$component->setValues(0);
+					});
+
+					$data->components[0]->setValues($data->unitary_value);
```
[the remark this answers is not in the document]
Comment (JoaoPedro0000) [reply Services/Products/CompositeService.php:38]: Adicionei sim o comentário para explicar essa ação, porém com o seguinte texto: 
`/* This was necessary due to the fact that the customer can choose that the value of the kit is not just the sum of the internal products (components). This value can be defined manually and must be used to correctly calculate shipping insurance, this way the entire insurance value is passed within the first product */`

File: Services/Products/ProductsService.php (hunk 3 above)
Comment (JoaoPedro0000) [reply Services/Products/ProductsService.php:90]: Vou criar uma task para isso, mas essa função é utilizada e tem sentido, pelo que eu vi quando o usuário marcar a opção de não assegurar o valor, o payload irá passar por essa função e será removido o `unitary_value`  dos produtos, dessa forma ficando sem o valor assegurado, mas vai ser necessário refatorar isso e deixar essa questão mais explicita

File: Services/Products/ProductsService.php (hunk 4 above)
Comment (JoaoPedro0000) [reply Services/Products/ProductsService.php:169]: Isso seria um outro plugin, não entrei a fundo nessa outra classe, vou criar uma tarefa para validar se isso fica ou será removido do código

File: Services/Products/ProductsService.php (hunk 5 above)
Comment (JoaoPedro0000) [reply Services/Products/ProductsService.php:217]: Essa função `normalize()` é utilizada para todos os tipos de produtos, simples, composto, componente, qualquer tipo de produto pode ser passado ai, o objetivo dela é padronizar o objeto produto para um "formato" que julgamos mais fácil de lidar e com as informações que precisamos

File: Services/Products/ProductsService.php (hunk 6 above)
Comment (JoaoPedro0000) [reply Services/Products/ProductsService.php:286]: é

File: Services/Products/ProductsService.php (hunk 3 above)
Comment (tsantos8080) [reply Services/Products/ProductsService.php:70]: Na verdade, não, porquê no carrinho `products.*.unitary_value` e `products.*.insurance_value` são ignorados. Só é considerado o valor em `options.insurance_value`. Testei isso aqui e vou deixar a documentação também como referência:
> O campo `options.insurance_value` deve conter o valor de seguro do envio, que deve corresponder ao valor dos itens/produtos enviados e deverá bater com o valor da NF.

https://docs.melhorenvio.com.br/reference/inserir-fretes-no-carrinho

Comment (JoaoPedro0000) [reply Services/Products/ProductsService.php:70]: No caso ali é removido na hora da cotação e não na inserção ao carrinho, na hora da cotação se remover os valores do `products.*.unitary_value` ele não vai ter nenhum valor de seguro.

**Review by viniciustessmann — APPROVED** [review]
